x-pack/plugins/apm/server/lib/metrics/by_agent/shared/memory/index.ts

```diff
@@ -43,36 +46,93 @@ const chartBase: ChartBase = {
   series,
 };
 
-export const percentMemoryUsedScript = {
+export const percentSystemMemoryUsedScript = {
   lang: 'expression',
   source: `1 - doc['${METRIC_SYSTEM_FREE_MEMORY}'] / doc['${METRIC_SYSTEM_TOTAL_MEMORY}']`,
 };
 
+export const percentCgroupMemoryUsedScript = {
+  lang: 'painless',
+  source: `
```

Comment:
Are we 100% sure we cannot calculate this without painless?

Reply:
I'm not sure about the various scenarios, but I'm wondering if we need to support different scenarios per bucket, or only per series. E.g., are we expected to show a series that uses METRIC_CGROUP_MEMORY_LIMIT_BYTES in one bucket, METRIC_SYSTEM_TOTAL_MEMORY in another, and METRIC_CGROUP_MEMORY_STATS_INACTIVE_FILE_BYTES in another bucket?

If this is not a requirement, we could also consider using expressions, and running three separate (parallel) searches, guarded by `exists` queries. This would make the scripts much simpler, it would remove the need for us to think about mapping issues, and the queries should terminate fast in most scenarios if they are not relevant for our usage.

Reply:
We could also gather sums from ES, and calculate the actual values in ES (not sure if that is feasible).

Reply:
Example queries:

cgroup memory metrics:
```
{
  query: {
    bool: {
      filter: [
        {
          exists: {
            field: 'system.process.cgroup.memory.mem.usage.bytes'
          }
        },
        {
          exists: {
            field: 'system.process.cgroup.memory.mem.limit.bytes'
          }
        },
        {
          exists: {
            field: 'system.process.cgroup.memory.stats.inactive_file.bytes'
          }
        }
      ]
    }
  },
  aggs: {
    memoryUsedAvg: {
      avg: {
        script: {
          lang: 'expression',
          source: `1 - (doc['system.process.cgroup.memory.mem.usage.bytes'] - doc['system.process.cgroup.memory.stats.inactive_file.bytes']) / doc['system.process.cgroup.memory.mem.limit.bytes']
        }
      }
    }
  }
}
```

fallback metrics:
```
{
  query: {
    bool: {
      filter: [
        {
          exists: {
            field: 'system.memory.total'
          }
        },
        {
          exists: {
            field: 'system.memory.actual.free'
          }
        }
      ]
    }
  },
  aggs: {
    memoryUsedAvg: {
      avg: {
        script: {
          lang: 'expression',
          source: `1 - (doc['system.memory.free'] / doc['system.memory.total']`
        }
      }
    }
  }
}
```

I've left out unbounded memory checks, as @eyalkoren mentions the following:

> The proposed implementation is correct. The only change related in the agent spec is that UI should not see the special unlimited values as agents shouldn't send them, but I thought it couldn't harm do this check in UI anyways.

Reply:
Thanks for the suggestion @dgieselaar, but I didn't add the filter below in the current query, because there's a scenario where the `cgroup.memory.mem.usage` is available, but the `cgroup.memory.mem.limit` is not, and in that case the calculation must be `(cgroup.memory.mem.usage - cgroup.memory.stats.inactive_file) / system.memory.total`

```js
{
          exists: {
            field: 'system.process.cgroup.memory.mem.limit.bytes'
          }
        },
```

Reply:
In that case you could add a third query:

```
{
  query: {
    bool: {
      filter: [
        {
          exists: {
            field: 'system.process.cgroup.memory.mem.usage.bytes'
          }
        },
        {
          exists: {
            field: 'system.process.cgroup.memory.stats.inactive_file.bytes'
          }
        }
      ],
      must_not: [
        {
          exists: {
            field: 'system.process.cgroup.memory.mem.limit.bytes'
          }
        }
      ]
    }
  },
  aggs: {
    memoryUsedAvg: {
      avg: {
        script: {
          lang: 'expression',
          source: `1 - (doc['system.process.cgroup.memory.mem.usage.bytes'] - doc['system.process.cgroup.memory.stats.inactive_file.bytes']) / doc['system.memory.total']
        }
      }
    }
  }
}
```

but then it's maybe not much of an improvement

Reply:
> Are we 100% sure we cannot calculate this without painless?

I'd like to move the calculation to javascript but since the max and average aggregations are based on the calculation I didn't find a way to do so. 😣

Reply:
I'll see if I can simplify it now that the API tests are ready,

Reply:
I've tested some stuff, like having multiple queries, and I still think the painless script is more clear.